proposals/public-suffix.md

```diff
@@ -0,0 +1,824 @@
+or whether to try a navigation, when a user enters a domain-like string in the
+url bar. In such instances, a PSL lookup is made and:
+
+* If the domain has a known eTLD, attempt to navigate.
+* If the domain has an unknown eTLD, use a search engine.
+
+#### 5. Site-specific data
+
+Extensions sometimes need to associate data with a hostname's "site", which may be:
+
+* a registrable domain (i.e. with a known eTLD)
+* an IP address
+* an intranet hostname with a non-public (i.e. unknown) eTLD, or without any suffix
+
+Examples of this kind of data include cookies and password autofill.
+
+### Known Consumers
+
+Mozilla intends to make use of this API in its [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/)
+extension to allow users to create containers that isolate organizational base domains:
+[PR #2352](https://github.com/mozilla/multi-account-containers/pull/2352).
+Currently, users must manually curate their containers to capture all anticipated subdomains
+of an organization's base domain, and this manual approach is cumbersome and error-prone.
+
+In addition to this, an analysis has been done of the relevant use cases of a selection of
+other extensions that currently roll their own solutions in order to parse the PSL, and the
+use cases are set out in the [Use Cases](#use-cases) section. Although the extension
+authors have not been consulted directly, it is assumed that they would stand to benefit
+from this API due to the reasons stated in the [Objective](#objective) section.
+
+| Extension | Relevant use cases | Description |
+|-----------|--------------------|-------------|
+| [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) | #1, #2 | isolate requests into containers by organizational base domain, view all domains grouped by organization |
+| [uBlock Origin](https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/) | #1, #2, #3 | create organization/3rd-party request-blocking rules, view request blocking status by organization, create blocking rules for specifically 3rd-party requests |
+| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
+| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
+| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any eTLD |
+
+## Specification
+
+### Schema
+
+A new API `publicSuffix` is added as follows:
+
+```ts
+namespace publicSuffix {
+
+
+  // METHODS
+
+  // Determines if the given hostname is itself a known eTLD (i.e. in the PSL).
+  export function isKnownSuffix(
+    hostname: string,
```

Comment:
```suggestion
  // Determines if the given suffix is itself a known eTLD (i.e. in the PSL).
  export function isKnownSuffix(
    suffix: string,
```

Nit: The input is not a hostname but a suffix.

Reply:
I am using the term "hostname" in the sense of the [hostname property](https://developer.mozilla.org/en-US/docs/Web/API/URL/hostname) of javascript's URL class:

```js
const hostname = new URL("https://co.uk").hostname;
const isKnownSuffix = publicSuffix.isKnownSuffix(hostname);
```

The input parameter may be any hostname, and if it is anything other than a known suffix (e.g. an IP address, or an eTLD+1), then the method returns `false`.

In section "Behaviours: 4. Invalid hostname" I state:

> This API's methods should throw an error if a hostname passed as an input parameter
> [contains invalid characters]

I intended that same error-throwing behaviour to apply to this method too, so referring to the input parameter as a "hostname" is a way of conveying that intent.